### PR TITLE
feat: add theming support with themed color tokens

### PR DESCRIPTION
**Pull request by matebek-tw**
## 🖼 Context

We decided to add theming support as part of our next major release to be able to resolve some of the stylistic issues resulted by the new accessible color tokens on product pages.

## 🚀 Changes

- Add `ThemeProvider`
- Add `useTheme` hook and `withTheme` HOC to be able to access the theme variable
- Use the themed color tokens instead of the base tokens and assign them to the `:root` element

## 🤔 Considerations

- Theming just won't work on IE11 and below, no matter that if we use the `postcss-custom-properties` plugin, because the fallback values will be added to the stylesheet in build time, and in case the browser does not support variables, it will use the fallback value, which won't be re-evaluated in run-time.
- Injecting the `className` only works if the `children` of the `ThemeProvider` is a valid React element, as we have to use `React.cloneElement` to achieve this - adding a wapping `<div />` would break layouts for the component, finding the first valid element is not as simple as it sounds, I have no better idea at the moment, than just show a warning to the developers.

## ✅ Checklist

- [x] Make PR title meaningful and follow [the commit lint format](https://github.com/transferwise/neptune-web/blob/master/CONTRIBUTING.md#versioning-and-commit-lint) (it will appear in the changelog)
- [x] Changes are tested and all tests pass
- [ ] Changes meet [accessibility standards](https://github.com/transferwise/neptune-web/blob/main/ACCESSIBILITY.md) and there are no violations in the console
- [ ] Changes work in all supported browsers (don't forget IE11)
- [ ] You've updated the documentation if necessary

### CODE REVIEW

File: packages/components/src/common/hocs/theme/withTheme.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { ThemeType } from '../..';
```
Comment (dozortsev) [inline packages/components/src/common/hocs/theme/withTheme.tsx:1]: this file seems doesn't have any JSX so file could have `.ts` extension

File: packages/components/src/test-utils/index.js [hunk 2]
```diff
@@ -13,12 +13,12 @@ import en from '../i18n/en.json';
...
 function customRender(
   ui,
-  { locale = DEFAULT_LOCALE, messages = en, direction, ...renderOptions } = {},
+  { locale = DEFAULT_LOCALE, messages = en, theme = Theme.LIGHT, direction, ...renderOptions } = {},
 ) {
   // eslint-disable-next-line react/prop-types
   var Wrapper = ({ children }) => {
     return (
-      <Provider i18n={{ locale, messages }} direction={direction}>
+      <Provider i18n={{ locale, messages }} theme={theme} direction={direction}>
```
Comment (dozortsev) [inline packages/components/src/test-utils/index.js:21]: but atm there no `theme` prop in `Provider` component

File: packages/components/.storybook/preview.js [hunk 3]
```diff
@@ -84,10 +85,15 @@ const DirectionDecorator = (storyFn) => {
 
 const ProviderDecorator = (storyFn) => {
   const locale = select('locale', severalExamplesOfSupportedLocales, DEFAULT_LOCALE);
+  const theme = select('theme', Object.values(Theme));
```
Comment (dozortsev) [inline packages/components/.storybook/preview.js:88]: we don't need this as we don't launch theming yet

File: packages/components/src/utilities/styleInjector/StyleInjector.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { Children, cloneElement, isValidElement } from 'react';
+
+export const StyleInjector: React.FC<{ className?: string }> = ({ children, className }) => {
```
Comment (dozortsev) [inline packages/components/src/utilities/styleInjector/StyleInjector.tsx:3]: I would actually keep this logic inside ThemeProvider rather than abstract this in separate component, wdyt?

File: packages/components/src/utilities/styleInjector/StyleInjector.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  const styledChildren = Children.map(children, (child) => {
+    if (!isValidElement(child)) {
```
Comment (dozortsev) [inline packages/components/src/utilities/styleInjector/StyleInjector.tsx:5]: just discovered one more bug (but perhaps you were already aware about that), it doesn't add CSS class to React component that have "legit" (e.g `div` or `button` or `span`) HTML element at root

File: packages/components/src/provider/theme/ThemeProvider.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
+import { render, screen } from '@testing-library/react';
+
+import { Theme } from '../../common/theme';
+
+import { ThemeProvider } from './ThemeProvider';
+
+describe('ThemeProvider', () => {
```
Comment (dozortsev) [inline packages/components/src/provider/theme/ThemeProvider.spec.tsx:7]: we need add one more case here, when children is React component with real HTML element, you could create custom component here or use one of existing (e.g Button, Modal, Accordion ...)

File: packages/components/src/utilities/styleInjector/StyleInjector.tsx (hunk 5 above)
Comment (matebek-tw) [reply packages/components/src/utilities/styleInjector/StyleInjector.tsx]: Added a test to cover this case in https://github.com/transferwise/neptune-web/pull/1200/commits/fe19f15d33919e6630145bb5797eea148ed8ce0f. This should work, so probably something else causing this problem.

File: packages/components/src/snackbar/SnackbarProvider.js [hunk 7]
```diff
@@ -45,7 +45,7 @@ SnackbarProvider.propTypes = {
 };
 
 SnackbarProvider.defaultProps = {
-  timeout: 4500,
+  timeout: 450000,
```
Comment (helendurranttw) [inline packages/components/src/snackbar/SnackbarProvider.js:48]: Why was this changed?

File: packages/components/src/tabs/Tabs.less [hunk 8]
```diff
@@ -74,7 +74,6 @@
   &__panel {
     &--selected {
       padding: var(--size-8);
-      border: 1px solid @color-base-navy-light;
```
Comment (helendurranttw) [inline packages/components/src/tabs/Tabs.less:77]: Why was this removed?

Comment (dozortsev) [reply packages/components/src/tabs/Tabs.less:77]: I recall that it wasn't making any affect so I removed since we not long ship base colors

you can try compare [main](https://transferwise.github.io/neptune-web/components/content/Tabs) with [this one](https://transferwise.github.io/neptune-web/branch/add-theme-provider/components/content/Tabs)

File: packages/components/src/snackbar/SnackbarProvider.js (hunk 7 above)
Comment (dozortsev) [reply packages/components/src/snackbar/SnackbarProvider.js:48]: opps, this needs to reverted

```suggestion
  timeout: 4500,
```

File: .github/workflows/ci-cd.yml [hunk 9]
```diff
@@ -289,7 +289,7 @@ jobs:
     name: Publish beta packages
     if: ${{ github.ref != 'refs/heads/main' }}
     runs-on: ubuntu-latest
-    needs: test
+    needs: setup
```
Comment (dozortsev) [inline .github/workflows/ci-cd.yml:292]:
```suggestion
    needs: test
```

File: packages/css/src/less/buttons.less [hunk 10]
```diff
@@ -93,21 +93,22 @@
...
       var(--color-background-screen);
-      var(--color-base-white-light);
+      #fff;
     );
   }
 }
 
 // @DEPRECATED:.btn-link (use Button component with type Accent / priority Tertiary)
 .btn-link,
+.btn-info,
```
Comment (helendurranttw) [inline packages/css/src/less/buttons.less:111]: ?

File: packages/css/src/less/buttons.less [hunk 11]
```diff
@@ -143,36 +144,6 @@
   }
 }
 
-// @DEPRECATED:.btn-success (use Button component with type Positive)
```
Comment (helendurranttw) [inline packages/css/src/less/buttons.less:146]: Why did this move?

File: packages/css/src/less/buttons.less [hunk 12]
```diff
@@ -220,19 +221,37 @@
...
       var(--color-interactive-negative-hover);
       var(--color-interactive-negative-active);
       var(--color-background-screen);
-      var(--color-base-white-light);
+      #fff;
     );
   }
 }
 
+// @TODO Deprecate / Remove
```
Comment (helendurranttw) [inline packages/css/src/less/buttons.less:237]: Why did this move?

File: packages/css/src/less/core/_scaffolding.less [hunk 13]
```diff
@@ -54,6 +54,19 @@ a,
   text-decoration: underline;
   background-color: transparent; // Normalize 8.0.1
 
+  .bg-primary & {
+    color: @color-navy-content-accent;
```
Comment (helendurranttw) [inline packages/css/src/less/core/_scaffolding.less:58]: Should this not be interactive?

File: packages/css/src/less/core/_scaffolding.less [hunk 14]
```diff
@@ -70,7 +83,7 @@ a,
   }
 
   &.secondary {
-    color: @color-base-navy-light;
+    color: var(--color-content-positive);
```
Comment (helendurranttw) [inline packages/css/src/less/core/_scaffolding.less:86]: Positive?

File: packages/css/src/less/core/_typography.less [hunk 15]
```diff
@@ -225,7 +226,7 @@ blockquote {
 
   footer {
     display: block;
-    color: @color-base-grey-light;
+    color: #fff;
```
Comment (helendurranttw) [inline packages/css/src/less/core/_typography.less:229]: Grey to white?

File: packages/css/src/less/footer.less [hunk 16]
```diff
@@ -60,50 +60,51 @@
     text-decoration: none;
 
     svg {
-      fill: var(--color-interactive-secondary);
+      fill: var(--color-content-primary);
```
Comment (helendurranttw) [inline packages/css/src/less/footer.less:63]: Completely different?

File: packages/docs/components/Layout.js [hunk 17]
```diff
@@ -54,6 +54,7 @@ const Layout = ({ children, router: { pathname } }) => {
               </Link>
             </li>
           ))}
+        <li className="Nav__Group">Configs</li>
```
Comment (helendurranttw) [inline packages/docs/components/Layout.js:57]:
```suggestion
        <li className="Nav__Group">Config</li>
```

File: packages/css/src/less/neptune.bundle.less [hunk 18]
```diff
@@ -32,7 +31,6 @@
 @import './loaders.less';
 @import './media.less';
 @import './modals.less';
-@import './panels.less';
```
Comment (helendurranttw) [inline packages/css/src/less/neptune.bundle.less:35]: Why was this removed?

File: packages/css/src/props/legacy-custom-props.css [hunk 19]
```diff
@@ -1,65 +0,0 @@
-/*
```
Comment (helendurranttw) [inline packages/css/src/props/legacy-custom-props.css:1]: Why was this removed? This needs to be put back. We can tell people we'll be removing them, but unless you provide a simple migration strategy or give them time to update, you cannot simply remove props they may be depending on. At a minimum you need to let them know what you're removing. A simple search of codesearch shows 119 usages of `--spacer` alone.

File: packages/css/src/variables/neptune-tokens.less [hunk 20]
```diff
@@ -1,2 +1 @@
 @import '@transferwise/neptune-tokens/tokens.less';
-@import '@transferwise/neptune-tokens/colors-base.less';
```
Comment (helendurranttw) [inline packages/css/src/variables/neptune-tokens.less:2]: This is another breaking change. I realise the base colours in tokens may not be what they were before, but we should do our best to alias them or give people an upgrade guide. From what I see the tokens are mostly the same, except for @color-base-white, which we could add a manual alias for to @color-base-white-light.

File: packages/docs/public/static/assets/main.css [hunk 21]
```diff
@@ -1,3 +1,5 @@
+@import "@transferwise/neptune-tokens/themes/navy/tokens.css";
```
Comment (helendurranttw) [inline packages/docs/public/static/assets/main.css:1]: I don't think this is necessary

**Comment by helendurranttw** [comment]
This PR is way too big. Like the release PR you have open to be merged into main, it is very difficult to review, because of the breadth of changes. Many of the changes are unrelated to the original scope of the PR, and there are several important breaking changes in here which are not explained or called out in the PR description.

Please try and keep unrelated changes out of PRs, and try and break the changes up across PRs to make reviewing easier. Thank you!

File: packages/css/src/less/buttons.less (hunk 10 above)
Comment (dozortsev) [reply packages/css/src/less/buttons.less:111]: fixed in https://github.com/transferwise/neptune-web/pull/1231

File: packages/css/src/less/buttons.less (hunk 11 above)
Comment (dozortsev) [reply packages/css/src/less/buttons.less:146]: fixed in https://github.com/transferwise/neptune-web/pull/1231

File: packages/css/src/less/buttons.less (hunk 12 above)
Comment (dozortsev) [reply packages/css/src/less/buttons.less:237]: fixed in https://github.com/transferwise/neptune-web/pull/1231

File: packages/css/src/less/core/_typography.less (hunk 15 above)
Comment (dozortsev) [reply packages/css/src/less/core/_typography.less:229]: fixed in https://github.com/transferwise/neptune-web/pull/1231

File: packages/css/src/less/footer.less (hunk 16 above)
Comment (dozortsev) [reply packages/css/src/less/footer.less:63]: fixed in https://github.com/transferwise/neptune-web/pull/1231
